### PR TITLE
Equivalence injective type families

### PR DESCRIPTION
**Pull request by fredrik-bakke**
I was checking out Agdas `INJECTIVE` pragma. Here are some edits I made while exploring.

### CODE REVIEW

**Comment by fredrik-bakke** [comment]
I completely agree that we dont want to use this pragma, and this PR doesn't introduce its use. I was curious if we could make Agda infer things like the cardinality of a standard finite set with it, but found out that functionality doesn't yet exist.

File: src/foundation/injective-type-families.lagda.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+By [univalence](foundation-core.univalence.md), these two structures are
+equivalent, but more generally every equivalence injective type family must
+always be injective as a map.
+
+## Definition
+
+### Equivalence injective type families
+
+```agda
+is-equiv-injective : {l1 l2 : Level} {A : UU l1} → (A → UU l2) → UU (l1 ⊔ l2)
```
Comment (EgbertRijke) [inline src/foundation/injective-type-families.lagda.md:44]: May I suggest to call this `is-equivalence-injective` because `is-equiv` is such a well established shorthand for the assertion that something is an equivalence?

File: src/foundation/injective-type-families.lagda.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+A type family `P : A → UU` is
+{{#concept "injective" Disambiguation="type family"}} if it is
+[injective as a map](foundation-core.injective-maps.md). However, we can also
+consider injectivity with respect to
+[equivalences of types](foundation-core.equivalences.md), which we dub
+{{#concept "equivalence injectivity" Disambiguation="type family" Agda=is-equiv-injective}}.
+By [univalence](foundation-core.univalence.md), these two structures are
+equivalent, but more generally every equivalence injective type family must
+always be injective as a map.
```
Comment (EgbertRijke) [inline src/foundation/injective-type-families.lagda.md:37]: There could be a more general notion of "observationally injective maps" where we have a map `f : A -> B` between two types that are both equipped with an identity system.

**Comment by EgbertRijke** [comment]
This file is compiling and it looks ready for review. I would be happy to give it my approval and merge it after the open conversations are resolved.

**Comment by fredrik-bakke** [comment]
> This PR is compiling and it looks ready for review. I would be happy to give it my approval and merge it after the open conversations are resolved.
> 
> In particular, there are lots of edits in this PR that would be useful to merge into the library, and if you have further developments of equivalence injective type families in mind they could easily happen in a subsequent pull request.

Thanks, I am not planning on adding anything else. I was just unsure of the naming and wanted to review the code myself before marking it ready for review, but seems you helped expedite that process 😅

File: src/foundation/injective-type-families.lagda.md (hunk 2 above)
Comment (fredrik-bakke) [reply src/foundation/injective-type-families.lagda.md]: I'm deferring this to future work

File: src/foundation/injective-type-families.lagda.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
+# Injective type families
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/injective-type-families.lagda.md]: Thanks for pointing that out!

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/injective-type-families.lagda.md]: Are you alright with this wording:

````md
We can say a type family `P : A → 𝒰` is
{{#concept "injective" Disambiguation="type family"}} if it is
[injective as a map](foundation-core.injective-maps.md). However, we can also
consider injectivity with respect to
````

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/foundation/injective-type-families.lagda.md]: The wording is clear, but I think it is questionable whether we should introduce a concept of injective type families in this way at all.

The point is that type families, despite them being defined as maps into the universe, are objects. For objects injectivity already has a meaning, and it is not the one you propose here. When I had that file about extensions of spans/span diagrams you rightfully pointed out to me that "extension" already has a meaning for morphisms, and it was not the one that was given by me at the time. So I changed it to concatenation. But I think it is exactly the same situation here. Injectivity already has a meaning for objects, and type families are objects, so we should be careful giving the "injective map" meaning to "injective type family".

I also think that we don't really need to introduce this notion of injective type family at all. We can just use the predicate `is-injective` from `foundation.injective-maps` whenever we wish to express that `B x = B y -> x = y` for all `x` and `y`.

My proposal is therefore that we only introduce "equivalence injective type families", which is what I believe you originally had in your pull request anyway, and skip introducing 'injective type family".

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/foundation/injective-type-families.lagda.md]: The terminology "equivalence injective type family" is still a bit subobtimal, because it is so close to suggesting the definition of injective type family that I'm not really fond of. So I've been thinking about whether there could be more interesting concepts lurking under here.

The family of maps in equivalence injective type families are converse to the transport maps `tr B : (x = y) -> (B x <~> B y)`. The main example of equivalence injective type families is the family `Fin`, but for `Fin` something more holds: this map has a section. So how about abolishing any suggestion of "injective type family" altogether, and call these type families `transport-split`? 

In other words, a type family `B` would be called `transport-split` if the map `tr B : (x = y) -> (B x <~> B y)` has a section for each `x y : A`. In particular, if a type family is transport split, we have a family of maps `(B x <~> B y) -> (x = y)` indexed by `x y : A`, and moreover, this family of maps sends `id-equiv (B x)` to `refl`.

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/injective-type-families.lagda.md]: I guess in the end, part of this discussion boils down to the fact that "injectivity" is just an ill-behaved notion. But I agree that what you put forward is a much more interesting concept to focus on. Perhaps we can find applications where we are able to avoid univalence using it? For instance, if `Fin` is transport split, shouldn't it satisfy a fragment of equivalence induction without relying on univalence? I can write up a file about this as soon as I find the time.

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/foundation/injective-type-families.lagda.md]: Certainly, injectivity for maps is also ill-behaved.

However, injectivity for objects is not so ill behaved. And I think my point stands (independently of whether injectivity for maps is ill-behaved) that we shouldn't call any kind of objects "injective" if that does not refers to the well-established categorical notion of injectivity for objects.

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/foundation/injective-type-families.lagda.md]: > I guess in the end, part of this discussion boils down to the fact that "injectivity" is just an ill-behaved notion. But I agree that what you put forward is a much more interesting concept to focus on. Perhaps we can find applications where we are able to avoid univalence using it? For instance, if `Fin` is transport split, shouldn't it satisfy a fragment of equivalence induction without relying on univalence? I can write up a file about this as soon as I find the time.

This would be very interesting! Let me know if you find anything out about it!

[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/foundation/injective-type-families.lagda.md]: Oh, sure, sorry for not addressing the injectivity of objects thing. I am not trying to argue against it, and I am on board about trying to find different terminology.

**Comment by fredrik-bakke** [comment]
If it's okay with you, may I continue my work on transport-split type families in this PR? It seems reasonable to keep it collected, and I don't expect this PR to be particularly large regardless.

**Comment by EgbertRijke** [comment]
> ~If it's okay with you, may I continue my work on transport-split type families in this PR? It seems reasonable to keep it collected, and I don't expect this PR to be particularly large regardless.~
> 
> EDIT: Never mind, I'll just make a new branch.

I was actually expecting that you would change the notion of equivalence injective type family to transport-split type family. Do you still want to go through with this PR as it is?

**Comment by fredrik-bakke** [comment]
> > ~If it's okay with you, may I continue my work on transport-split type families in this PR? It seems reasonable to keep it collected, and I don't expect this PR to be particularly large regardless.~
> > 
> > EDIT: Never mind, I'll just make a new branch.
> 
> I was actually expecting that you would change the notion of equivalence injective type family to transport-split type family. Do you still want to go through with this PR as it is?

I think equivalence injective type families deserve a spot on equal footing to injective maps. I was thinking to develop the other notions alongside it.

**Comment by EgbertRijke** [comment]
Ah ok, I thought we came to a different conclusions, i.e., that objects should not be called injective in this way even if they happen to be defined as some maps.

To me, the notion of injectivity for maps is acceptible because it is a well established notion about maps, but to call an object injective in a way that doesn't refer to the usual notion of injective object in any way sounds wrong and I don't know why it should have equal footing to the well established notion of injective map.

**Comment by fredrik-bakke** [comment]
Okay. Then how about `is-equal-index-is-equivalent-fibers`? If for nothing else, the notion is already used in some places, and serve as a way to give _some_ computational content to "injectivity of maps" for type families

**Comment by EgbertRijke** [comment]
Hm... my apologies I think I went a bit too hard in my previous message. It is totally true that the notion has been used previously, and that it is a useful thing to isolate. Perhaps "equivalence injective type family" is just the best name for it.

I think I can live with it:P Apologies for saying that it is wrong.

**Comment by EgbertRijke** [comment]
I checked this PR once more and it looks good now. I'll merge it. Thanks for your good work Fredrik, and thanks for bearing with me.

**Comment by fredrik-bakke** [comment]
No worries! Keep posted for the sequel 😁
